### DISCUSSION TITLE
Ring token refresh

### DISCUSSION THREAD

**Discussion by DaveDixon** [tsightler/ring-mqtt#390] [General Discussion]
I've had Ring working in HA off and on for a long time. Every time something changed (2FA, for example) I let it stay broken for a while then got it going again. I just got ring-mqtt going again ten days ago when I upgraded HA to 2022.6.6 (core). It worked fine for 8 or 9 days, then started failing on a refresh token. On investigation, I see that I have an artifact homebridge service running that, as I recall, was supposed to handle the token refresh. Is that still required? Or are these things colliding?

**Comment by tsightler** [comment]
> artifact homebridge service

I have no idea what this is, but ring-mqtt maintains its own refresh tokens and certainly doesn't depend on anything external to do so.  I wouldn't expect any conflicts with anything else as each ring-mqtt instance maintains a unique systemId with its own refresh tokens (same as having two devices with the Ring app installed using the same account).

You didn't state anything about how you have ring-mqtt installed, but if you are using Docker then you must include persistent storage for the refresh token.   Sometimes such issues are caused by outages on Ring servers, for example, there was a pretty significant issue on June 21st, which caused refresh token failures, but I haven't seen any issues since then.

**Reply by DaveDixon** [reply]
Thanks for the quick reply!

I thought I'd mentioned in my post that I'm using HA core - so not Docker.

The error message from ring-mqtt says this:

  Refresh token is not valid.  Unable to authenticate with Ring servers.  See https://github.com/dgreif/ring/wiki/Refresh-Tokens

On the page https://github.com/dgreif/ring/wiki/Refresh-Tokens it talks about using homebridge-ring to refresh tokens. My recollection is that right after Ring implemented 2FA this was the only way to keep the Ring integration going. If mqtt-ring takes care of that now, then obviously I can disable homebridge-ring.

**Reply by tsightler** [reply]
You did mention HA core, but that doesn't preclude using the ring-mqtt Docker install, as a matter of fact, that is the recommended setup.  Basically, you can pretty much mix-and-match HA and ring-mqtt install methods, with the exception that the addon install method requires HA supervised.  Since you say you are using HA core, that means you could use Docker or Standard install methods.

The refresh token page mentions homebridge-ring because dgreif is the primary maintainer of both homebridge-ring and ring-client-api.  Basically ring-client-api is the underlying client library used by both homebridge-ring and ring-mqtt.  The ring-mqtt code maintains its own refresh tokens and has no relation to homebridge-ring.

Since you indicate that you are not using the Docker install, then I'll assume you are using the standard install.  You should verify that there is a ring-state.json file being created in the directory where ring-mqtt is installed, and it should contain the most recent token.  You can always just re-authenticate, but, in normal operation, new tokens are acquired approximately every hour and saved in this file for future use.
